lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,53 +40,147 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(text, str) or not isinstance(end_of_word, str):
+        return None
+    if not isinstance(start_of_word, str) and start_of_word is not None:
+        return None
+
+    dict_freq = {}
+    words_list = text.split()
+    for word in words_list:
+        prepared_word = prepare_word(word, start_of_word, end_of_word)
+        if prepared_word is None:
+            return None
+        if prepared_word not in dict_freq:
+            dict_freq[prepared_word] = words_list.count(word)
+
+    return dict_freq
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_dict = {}
+
+    for word in word_frequencies:
+        for index in range(len(word) - 1):
+            pair = (word[index], word[index + 1])
+            if pair not in pair_dict:
+                pair_dict[pair] = 0
+            pair_dict[pair] += word_frequencies[word]
+
+    return pair_dict
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+
+    word_freq_updated = word_frequencies.copy()
+
+    for word in word_frequencies:
+        new_word = []
+        if pair[0] in word and pair[1] in word:
+            for index, token in enumerate(word):
+                if token == pair[1] and word[index - 1] == pair[0]:
```

Comment:
I don't understand why we are passing here

Reply:
It helps to prevent adding pair[1] to new_word two times. It checks if the token we are checking is pair[1] and the previous one is pair[0]. If those tokens are like that, that means that both of them are already in the new_word, because I append both tokens from the pair if the token is pair[0] and the next one is pair[1]. So in order to not add pair[1] again to new_word we pass in here.